packages/flutter_tools/lib/src/device.dart

```diff
@@ -99,6 +113,21 @@ abstract class Device {
   String toString() => '$runtimeType $id';
 }
 
+/// Read the log for a particular device. Subclasses must implement `hashCode`
+/// and `operator ==` so that log readers that read from the same location can be
+/// de-duped. For example, two Android devices will both try and log using
+/// `adb logcat`; we don't want to display two identical log streams.
+abstract class DeviceLogReader {
+  String get name;
+
+  Future<int> logs({bool clear: false});
+
+  int get hashCode => name.hashCode;
```

Comment:
did you mean to have this one be implemented? It means that the analyzer won't complain if subclasses don't override it.

If you did, then remove the one on the android side since it just does the same thing.